### ISSUE TITLE
Handle DNS resolution failure

### ISSUE THREAD

**Issue by 2Max4** [alessandromaggio/pythonping#28]
Now and then I am getting the following error:

Traceback (most recent call last):
  File "network_test.py", line 164, in <module>
    test_result = get_ping_as_df(url)
  File "network_test.py", line 125, in get_ping_as_df
    my_ping = ping(ping_url)
  File "lib\site-packages\pythonping\__init__.py", line 52, in ping
    comm = executor.Communicator(target, provider, timeout, socket_options=options, verbose=verbose, output=out)
  File "C:\lib\site-packages\pythonping\executor.py", line 224, in __init__
    self.socket = network.Socket(target, 'icmp', source=None, options=socket_options)
  File "lib\site-packages\pythonping\network.py", line 21, in __init__
    self.destination = socket.gethostbyname(destination)
socket.gaierror: [Errno 11001] getaddrinfo failed
Exception ignored in: <function Socket.__del__ at 0x00000189769F9730>
Traceback (most recent call last):
  File "\lib\site-packages\pythonping\network.py", line 56, in __del__
    if self.socket:
AttributeError: 'Socket' object has no attribute 'socket'
Exception ignored in: <function Communicator.__del__ at 0x0000018976A009D8>
Traceback (most recent call last):
  File "\lib\site-packages\pythonping\executor.py", line 233, in __del__
    if self.socket:
AttributeError: 'Communicator' object has no attribute 'socket'

which causes my script for Network analysis to crash. Are there any known bugfixes yet?

Thanks for your support!

**Comment by alessandromaggio** [comment]
This error means you are not running the script as root. Are you all the time?

**Comment by darkwoolf10** [comment]
use sudo is fix, but it's not good. I think.

**Comment by alessandromaggio** [comment]
I don't think its either good or bad @darkwoolf10.
All the operating systems let normal applications use TCP and UDP freely (mostly), but ICMP runs directly on top of IP. So, to send ICMP packets, we need to have direct access to the IP level, and that is something very powerful.

Potentially, at that level, you can sniff the traffic of other applications, and even pretend to be another application. You can perform a man-in-the-middle attack for _all_ the application on the same host.

That's why you need to be root at the IP level, and to send ICMP packets we need that.

**Comment by 2Max4** [comment]
@alessandromaggio Yes I am running the script as sudo. Nevertheless this error occurs! What's interesting is that - if there was an error with permissioning - this error would occur directly after starting the script. This is not the case. There are already several ping checks made before this error occurs. 

Do you know if there is something like a "max-out" from the OS? Currently I am sending Ping checks every 5 seconds... maybe this is too much for the OS.

**Comment by alessandromaggio** [comment]
Weird. OS should not pose any limit on that, it either grants you access to the socket or not, but should not have any limitation once you do have access. I am worried that we are not doing some proper cleaning once the script ends, and re-opening the socket multiple times cause this to mess around.
Which OS are you running?

**Comment by 2Max4** [comment]
I am running this script on Windows 10 and Raspian.

I changed my Exception handling to a very generic one and since I have done this I do only get the Error: 

```
[Errno -3] Temporary failure in name resolution
Traceback (most recent call last):
  File "network_test.py", line 133, in get_ping_as_df
    my_ping = ping(ping_url)
  File "/usr/local/lib/python3.7/dist-packages/pythonping/__init__.py", line 52, in ping
    comm = executor.Communicator(target, provider, timeout, socket_options=options, verbose=verbose, output=out)
  File "/usr/local/lib/python3.7/dist-packages/pythonping/executor.py", line 224, in __init__
    self.socket = network.Socket(target, 'icmp', source=None, options=socket_options)
  File "/usr/local/lib/python3.7/dist-packages/pythonping/network.py", line 21, in __init__
    self.destination = socket.gethostbyname(destination)
socket.gaierror: [Errno -3] Temporary failure in name resolution

```

Which -from my understanding - says that there is no connection to the internet. 
Maybe the error from the socket is in a connection with the network error I've posted. Meaning that whenever I initially had no connection the socket error would occure.

Greetings

**Comment by alessandromaggio** [comment]
The issue is now clear, I have renamed it to "Handle DNS resolution failure".
This is a very important issue, yet easy to fix - we must include it in the next release.

**Comment by alessandromaggio** [comment]
Added a clearer explanation of why this error happen, raised as `RuntimeError`. As this is not an issue of pythonping itself, this is the best we can do. Change included in release 1.0.9, just published.
Thanks @2Max4 for raising this.

**Comment by bytetonight** [comment]
I believe this is a pythonping issue because of the way pythonping is handling the actual error, namely gaierror. Is it not possible to just raise that error, which can be handled in a users code, instead of a runtime error which just exits the program ?
The error only occurs if trying to ping a server that does not exist, or your connection is down. No need to crash everything then.
